teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,759 @@
+may fail to run, wasting resources on the cluster
+- Couples two Catalog Tasks together
+
+### Pipeline-level Sidecar
+
+Add support for Sidecars with the same lifespan as PipelineRuns, for example:
+
+```yaml
+kind: Pipeline
+spec:
+  params:
+  - name: image
+  - name: docker-host
+    default: $(sidecars.docker-daemon.host):2376
+  workspaces:
+  - name: source
+  - name: docker-tls-certs
+  results:
+  - name: image-digest
+    value: $(tasks.docker-build.results.IMAGE_DIGEST)
+  tasks:
+  - name: git-clone
+    ...
+  - name: docker-build
+    params:
+    - name: image
+    - name: docker-host
+    workspaces:
+    - name: source
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker build -f ./Dockerfile -t $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+      results:
+      - name: IMAGE_DIGEST
+    runAfter: ["git-clone"]
+  # This task uses the built image without first pushing it to a registry
+  - name: pytest
+    params:
+    - name: docker-host
+    - name: image
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker exec $(params.image) python -m pytest
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["docker-build"]
+  # This task pushes the image to the registry only if tests passed
+  - name: docker-push
+    when:
+    - input: "$(tasks.pytest.status)"
+      operator: in
+      values: ["Succeeded"]
+    params:
+    - name: image
+    - name: docker-host
+    taskSpec:
+      steps:
+      - image: docker.io/library/docker
+        script: docker push $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["pytest"]
+  sidecars:
+  - name: docker-daemon
+    image: docker:dind
+    securityContext:
+      privileged: true
+    ports:
+    - 2376
+    startupProbe:
+      tcpSocket:
+        port: 2376
+    workspaces:
+    - name: docker-tls-certs   
+```
+
+The sidecar would likely be implemented as a deployment with one pod plus a service.
+If the sidecar has readiness or startup probes, the PipelineRun would wait for them to succeed before starting the Pipeline Tasks.
+We could use a [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+to prevent other PipelineRuns from being able to use the sidecar.
+
+Pros:
+- Good separation of concerns: App developer creating build Pipeline doesn't have to worry about using the correct service accounts
+or kube credentials
+- Arguably better UX than using separate setup and teardown Tasks. Don't have to worry about resource leakage if teardown Task
+fails due to a misconfigured PipelineRun.
+
+Cons:
+- Sidecar configurations will likely be copy-pasted between Pipelines.
```

Comment:
Could be made "shareable" somehow (a new type, or a way to refer to a common definition from something, like form bundles, or …)

Reply:
hm can you elaborate on what you had in mind? I remember you weren't in favor of standalone/reusable sidecars so I'm surprised to see the suggestion of a new type. Hoping to keep most of the design discussion to a later PR though.